File: android/app/src/main/java/com/noop/analytics/CurrentHrv.kt
package com.noop.analytics

import com.noop.data.RrInterval

/**
 * Trailing-window "current HRV" — RMSSD over the most recent strap R-R rows, refreshed after each
 * successful sync. Separate from nightly `avgHrv` (sleep-window RMSSD fed into recovery); this is an
 * additive live readout only.
 *
 * Kotlin parity twin of `Packages/StrandAnalytics/.../CurrentHRV.swift`. Reuses [HrvAnalyzer] primitives only.
 */
object CurrentHrv {

    data class Snapshot(
        val rmssdMs: Double,
        val cleanBeats: Int,
        val coverage: Double,
        val computedAtUnix: Int,
    )

    /** Trailing window length (seconds) for the current HRV readout. */
    const val WINDOW_SECONDS: Int = HrvWindow.SECONDS

    /** Rows newer than this many seconds before `nowUnix` are treated as stale by the app-layer caller. */
    const val STALE_THRESHOLD_SECONDS: Int = 900

    /**
     * Derive a current HRV snapshot from R-R rows whose timestamps fall in
     * `[nowUnix - windowSeconds, nowUnix]`. Returns null when coverage fails the nightly RMSSD honesty
     * gate ([HrvAnalyzer.successiveDiffIsTrustworthy]) or when fewer than [HrvAnalyzer.MIN_BEATS] clean
     * beats survive.
     */
    fun derive(
        rows: List<RrInterval>,
        nowUnix: Int,
        windowSeconds: Int = WINDOW_SECONDS,
    ): Snapshot? {
        if (windowSeconds != HrvWindow.SECONDS) return null
        return deriveObservations(PhysiologyQuality.legacy(rows, "legacy-unscoped"), nowUnix)
    }

    /** The latest completed UTC window, never pooled with a previous sparse window. */
    fun deriveObservations(observations: List<PhysiologyQuality.IntervalObservation>, nowUnix: Int,
                           policy: HrvWindow.Policy = HrvWindow.Policy(), inputRevision: String = "unversioned"): Snapshot? {
        val result = HrvWindow.measure(HrvWindow.alignedStart(nowUnix) - HrvWindow.SECONDS,
            observations, policy = policy, inputRevision = inputRevision, computationMode = "causal")
        val rmssd = result.observedRMSSD ?: return null
        if (!result.measurementValid) return null
        return Snapshot(rmssd, Math.round(result.validIntervalFraction * result.originalIds.size).toInt(),
            result.observedTimeFraction, nowUnix)
    }
}

File: Packages/StrandAnalytics/Sources/StrandAnalytics/CurrentHRV.swift
import Foundation
import WhoopProtocol

/// Trailing-window "current HRV" — RMSSD over the most recent strap R-R rows, refreshed after each
/// successful sync. Separate from nightly `avgHrv` (sleep-window RMSSD fed into recovery); this is an
/// additive live readout only.
///
/// Swift parity twin of `android/.../analytics/CurrentHrv.kt`. Reuses `HRVAnalyzer` primitives only.
public enum CurrentHRV {

    public struct Snapshot: Equatable, Sendable {
        public let rmssdMs: Double
        public let cleanBeats: Int
        public let coverage: Double
        public let computedAtUnix: Int

        public init(rmssdMs: Double, cleanBeats: Int, coverage: Double, computedAtUnix: Int) {
            self.rmssdMs = rmssdMs
            self.cleanBeats = cleanBeats
            self.coverage = coverage
            self.computedAtUnix = computedAtUnix
        }
    }

    /// Trailing window length (seconds) for the current HRV readout.
    public static let windowSeconds = 30 * 60

    /// Rows newer than this many seconds before `nowUnix` are treated as stale by the app-layer caller.
    public static let staleThresholdSeconds = 900

    /// Derive a current HRV snapshot from R-R rows whose timestamps fall in
    /// `[nowUnix - windowSeconds, nowUnix]`. Returns nil when coverage fails the nightly RMSSD honesty
    /// gate (`successiveDiffIsTrustworthy`) or when fewer than `HRVAnalyzer.minBeats` clean beats survive.
    public static func derive(rows: [RRInterval], nowUnix: Int,
                              windowSeconds: Int = CurrentHRV.windowSeconds) -> Snapshot? {
        guard windowSeconds > 0 else { return nil }
        let windowStart = nowUnix - windowSeconds
        let seg = rows.filter { $0.ts >= windowStart && $0.ts <= nowUnix }
        guard !seg.isEmpty else { return nil }

        let ts = seg.map(\.ts)
        let rrMs = seg.map { Double($0.rrMs) }
        let coverage = HRVAnalyzer.rrCoverage(tsSec: ts, rrMs: rrMs)
        guard coverage > 0 else { return nil }

        // Same over-count refusal as `SleepStager.sessionAvgHRV` — `collapsed` pinned to `coverage` so
        // both over-count verdicts refuse without an extra sort (#1510).
        let verdict = HRVAnalyzer.classifyCoverage(coverage: coverage, collapsed: coverage)
        guard HRVAnalyzer.successiveDiffIsTrustworthy(verdict) else { return nil }

        let h = HRVAnalyzer.analyze(rawRR: rrMs)
        guard let rmssd = h.rmssd else { return nil }

        return Snapshot(rmssdMs: rmssd, cleanBeats: h.nClean, coverage: coverage, computedAtUnix: nowUnix)
    }
}
